frostsnap_core/src/device.rs

```diff
@@ -384,51 +413,141 @@ impl<S: NonceStreamSlot + core::fmt::Debug> FrostSigner<S> {
+                        ));
+                    }
+
+                    let (paired_secret_share, vrf_cert) = phase1
+                        .input_state
+                        .verify_receive_share_and_certify(
+                            &schnorr,
+                            &cert_scheme,
+                            (*my_index).into(),
+                            self.keypair(),
+                            &agg_input,
+                        )
+                        .map_err(|e| {
+                            Error::signer_invalid_message(
+                                &message,
+                                format!("Failed to verify and receive share: {e}"),
+                            )
+                        })?;
+
+                    let paired_secret_share = paired_secret_share.non_zero().ok_or_else(|| {
                         Error::signer_invalid_message(&message, "keygen produced a zero shared key")
                     })?;
 
-                    let phase2 = KeyGenPhase2 {
+                    self.tmp_keygen_phase2.insert(
+                        keygen_id,
+                        KeyGenPhase2 {
+                            keygen_id,
+                            device_to_share_index: phase1.device_to_share_index,
+                            paired_secret_share,
+                            agg_input,
+                            key_name: phase1.key_name.clone(),
+                            key_purpose: phase1.key_purpose,
+                            coordinator_public_key: phase1.coordinator_public_key,
+                        },
+                    );
+                    Ok(vec![DeviceSend::ToCoordinator(Box::new(
+                        DeviceToCoordinatorMessage::KeyGenCertify {
+                            keygen_id,
+                            vrf_cert,
+                        },
+                    ))])
+                }
+                self::Keygen::Check {
+                    keygen_id,
+                    certificate,
+                } => {
+                    let phase2 = self.tmp_keygen_phase2.remove(&keygen_id).ok_or_else(|| {
+                        Error::signer_invalid_message(
+                            &message,
+                            "no keygen state for provided keygen_id",
+                        )
+                    })?;
+
+                    // Reconstruct the certifier with all contributor public keys
+                    let cert_scheme = certpedpop::vrf_cert::VrfCertScheme::<Sha256>::new(
+                        crate::message::keygen::VRF_CERT_SCHEME_ID,
+                    );
+                    let contributor_keys: Vec<_> = phase2
+                        .device_to_share_index
+                        .keys()
+                        .map(|device_id| device_id.pubkey())
+                        .chain(core::iter::once(phase2.coordinator_public_key))
+                        .collect();
```

Comment:
Actually we only need the coordinator public key in here. The devices are part of agg_input.